pkg/instance/pool.go

```diff
@@ -0,0 +1,277 @@
+// The pool is assumed to have one default Runner (e.g. to be used for fuzzing), while a
+// dynamically controlled sub-pool might be reserved for the arbitrary Runners.
+type Pool[T Instance] struct {
+	BootErrors chan error
+
+	count      int
+	creator    CreateInstance[T]
+	defaultJob Runner[T]
+	jobs       chan Runner[T]
+
+	// The mutex serializes ReserveForRun() calls.
+	mu        sync.Mutex
+	instances []poolInstance[T]
+}
+
+func NewVMPool(pool *vm.Pool, def Runner[*vm.Instance]) *Pool[*vm.Instance] {
+	return NewPool(pool.Count(),
+		func(idx int) (*vm.Instance, error) {
+			return pool.Create(idx)
+		},
+		def)
+}
+
+func NewPool[T Instance](count int, creator CreateInstance[T], def Runner[T]) *Pool[T] {
+	instances := make([]poolInstance[T], count)
+	for i := 0; i < count; i++ {
+		instances[i].idx = i
+		instances[i].free(def)
+	}
+	return &Pool[T]{
+		BootErrors: make(chan error, 16),
+		count:      count,
+		creator:    creator,
+		defaultJob: def,
+		instances:  instances,
+		jobs:       make(chan Runner[T]),
+	}
+}
+
+func (p *Pool[T]) Loop(ctx context.Context) {
+	var wg sync.WaitGroup
+	wg.Add(p.count)
+	for i := 0; i < p.count; i++ {
+		i := i
+		go func() {
+			for ctx.Err() == nil {
+				p.runInstance(ctx, i)
+			}
+			wg.Done()
+		}()
+	}
+	wg.Wait()
+}
+
+func (p *Pool[T]) runInstance(ctx context.Context, idx int) {
+	obj := &p.instances[idx]
+	ctx, cancel := context.WithCancel(ctx)
+
+	log.Logf(2, "pool: booting instance %d", idx)
+
+	p.mu.Lock()
+	// Avoid races with ReserveForRun().
+	obj.reset(cancel)
+	p.mu.Unlock()
+
+	obj.status(StateBooting)
+	defer obj.status(StateOffline)
+
+	inst, err := p.creator(idx)
+	if err != nil {
+		p.BootErrors <- err
+		return
+	}
+	defer inst.Close()
+
+	obj.status(StateWaiting)
+	// The job and jobChan fields are subject to concurrent updates.
+	obj.mu.Lock()
+	job, jobChan := obj.job, obj.jobChan
+	obj.mu.Unlock()
+
+	if job == nil {
+		select {
+		case newJob := <-jobChan:
+			job = newJob
+		case newJob := <-obj.switchToJob:
+			job = newJob
+		case <-ctx.Done():
+			return
+		}
+	}
+
+	obj.status(StateRunning)
+	job(ctx, inst, obj.updateInfo)
+}
+
+func (p *Pool[T]) ReserveForRun(VMs int) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	if VMs > p.count {
+		panic("trying to reserve more VMs than present")
+	}
+
+	var free, reserved []*poolInstance[T]
+	for i := 0; i < p.count; i++ {
+		inst := &p.instances[i]
+		if inst.reserved() {
+			reserved = append(reserved, inst)
+		} else {
+			free = append(free, inst)
+		}
+	}
+
+	needReserve := VMs - len(reserved)
+	for i := 0; i < needReserve; i++ {
+		log.Logf(2, "pool: reserving instance %d", free[i].idx)
+		free[i].reserve(p.jobs)
+	}
+
+	needFree := len(reserved) - VMs
+	for i := 0; i < needFree; i++ {
+		log.Logf(2, "pool: releasing instance %d", reserved[i].idx)
+		reserved[i].free(p.defaultJob)
+	}
+}
+
+// Run() blocks until it has found an instance to execute job() and job() has finished.
```

Comment:
Remove "()". Go does not use () to mark functions in comments. Names are just names. There are also variables, arguments, types, packages, generic parameters, etc. Functions are not special.
https://tip.golang.org/doc/comment